Merge GPUTextureCopyView.arrayLayer into .origin.z

kainino0x opened the pull request:
Similar to what was done for GPUTextureDescriptor in #613, I just missed it.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/730.html" title="Last updated on May 8, 2020, 5:53 PM UTC (615fa8f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/730/be26167...kainino0x:615fa8f.html" title="Last updated on May 8, 2020, 5:53 PM UTC (615fa8f)">Diff</a>

Kangz commented:
I'd like to discuss the `GPUTextureDescriptor.arrayLayerCount` removal again. While trying to implement the change in Dawn, things got fairly confusing, and I believe @kvark got feedback from users of wgpu that they were confused too. I think we should revert the `arrayLayerCount` removal because it makes semantic less clear for texture creation, and consistency requires other changes like this one or #299 that make semantic less explicit.

Also none of the native APIs do this and instead they have an explicit array layer count:

 - [`MTLTextureDescriptor::arrayLength`](https://developer.apple.com/documentation/metal/mtltexturedescriptor/1515331-arraylength?language=objc)
 - [`VkImageCreateInfo::arrayLayers`](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#resources-images)
 - [`D3D12_RESOURCE_DESC::DepthOrArraySize`](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_resource_desc) arguable the closest, but it still explicitly calls out it is one or the other.

kainino0x commented:
Let's discuss it. If we revert then I'd like to do #520 instead.

Kangz commented:
After some offline discussion, looking copies available in all APIs, and remembering that 1D array textures exist (we should maybe have them in WebGPU), I'd like to retract my previous concern.

What we can do is better educate users and have nice representations of 2D array that looks like a cube and 1D array that looks like a plan.

kainino0x commented:
An example of spec complexity arising from separating arrayLayer from depth:

> * The number of slices of the extent (for 3D) or layers of the srcSubresource (for non-3D) must match the number of slices of the extent (for 3D) or layers of the dstSubresource (for non-3D)
> * If either of the calling command’s srcImage or dstImage parameters are of VkImageType VK_IMAGE_TYPE_3D, the baseArrayLayer and layerCount members of the corresponding subresource must be 0 and 1, respectively
> ...
> * If both srcImage and dstImage are of type VK_IMAGE_TYPE_2D then extent.depth must be 1
> * If the calling command’s srcImage is of type VK_IMAGE_TYPE_2D, and the dstImage is of type VK_IMAGE_TYPE_3D, then extent.depth must equal to the layerCount member of srcSubresource
> * If the calling command’s dstImage is of type VK_IMAGE_TYPE_2D, and the srcImage is of type VK_IMAGE_TYPE_3D, then extent.depth must equal to the layerCount member of dstSubresource

https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#VUID-VkImageCopy-srcImage-01790

kainino0x commented:
[editors call] Since this change is conceptually part of #613, which was approved by the group, merging without rehashing the discussion.

kainino0x commented:
Actually, now that we have landed copyTextureToTexture, this change needs more complete edits. While I was at it I fixed the subresource range set issue and added Origin3D.

kvark commented on spec/index.bs:
```diff
@@ -3197,11 +3201,28 @@ The following validation rules apply:
+        is {{GPUTextureDimension/"2d"}}:
+          - For each |arrayLayer| of the |copySize|.[=Extent3D/depth=] [=array layers=]
+            starting at |textureCopyView|.{{GPUTextureCopyView/origin}}.[=Origin3D/z=]:
+              - The [=subresource=] of |textureCopyView|.{{GPUTextureCopyView/texture}} at
```
I wonder if it's going to be cleaner to define a subresource in a function manner, i.e. `[$subresource$](texture, mipmapLevel, arrayLayer)`. This is just an idea, not something you have to fix here.

kvark commented on spec/index.bs:
```diff
@@ -3197,11 +3201,28 @@ The following validation rules apply:
+              - The [=subresource=] of |textureCopyView|.{{GPUTextureCopyView/texture}} at
+                [=mipmap level=] |textureCopyView|.{{GPUTextureCopyView/mipLevel}} and
+                [=array layer=] |arrayLayer|.
+      - If |textureCopyView|.{{GPUTextureCopyView/texture}}.{{GPUTexture/[[dimension]]}}
```
what about `1d` textures? maybe we should just say: `if 2d {..} else {..}`

kvark commented on spec/index.bs:
```diff
@@ -3646,6 +3667,24 @@ dictionary GPUOrigin3DDict {
 
+An <dfn dfn>Origin3D</dfn> is a {{GPUOrigin3D}}.
+[=Origin3D=] is a spec namespace for the following definitions:
+<!-- This is silly, but provides convenient syntax for the spec. -->
```
what's silly is that we allow both sequence and dictionary representation here, even though sequences are more heavy and can be trivially converted to a dictionary on the user side. This whole section wouldn't need to exist.

kainino0x replied:
Why are sequences more heavy? On the syntax side they're much lighter at least.

As you know I still want to accept only sequence here, but if there's a strong preference we can change it back to dictionary-only. FWIW at some point I do want to consider making it so different users of this type have different defaults (e.g. some require all 3, some require only 2 or 1).

kvark replied:
AFAIK, because sequence is a separate object in JS. For this reason we wanted to avoid them in `setVertexBuffers`, for example.

kainino0x replied:
A dictionary is also a separate object though. We would avoid a dictionary in setVertexBuffers too.

kvark replied:
Indeed! Would there be a difference when coming from WASM though?

kainino0x replied:
Yes actually, I keep forgetting about this. `sequence` allows a typed array, and actually I think any iterable (I just tried it with a Float32Array, and a generator `(function*() { ... })()` and those worked. Set didn't work but maybe it's supposed to.)

kvark replied:
So would it be correct to say that `sequence` is heavier for WASM case?

kainino0x replied:
I'd say it's slightly lighter. For WASM you can do any of:

- `{ x: HEAPF32[i], y: HEAPF32[i + 1], z: HEAPF32[i + 2] }` copy and construct dict
- `[ HEAPF32[i], HEAPF32[i + 1], HEAPF32[i  + 2] ]` copy and construct array
- `HEAPF32.slice(i, i + 3)` create new Float32Array + ArrayBuffer and memcpy
- `HEAPF32.subarray(i, i + 3)` create new Float32Array pointing to same ArrayBuffer
